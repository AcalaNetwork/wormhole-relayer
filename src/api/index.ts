export * from './relay';
export * from './testTimeout';
export * from './getVersion';
export * from './wormhole';
export * from './xcm';
export * from './homa';
export * from './euphrates';
export * from './health';
