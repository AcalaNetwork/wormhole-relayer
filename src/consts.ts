import {
  CHAIN_ID_ACALA,
  CHAIN_ID_KARURA,
} from '@certusone/wormhole-sdk';
import { ROUTER_TOKEN_INFO } from '@acala-network/asset-router/dist/consts';
import dotenv from 'dotenv';

dotenv.config({ path: '.env' });

const isTestnet = Number(process.env.TESTNET_MODE ?? 1);

export const enum ETH_RPC {
  LOCAL = 'http://localhost:8545',

  BSC = 'https://endpoints.omniatech.io/v1/bsc/mainnet/public',
  KARURA = 'https://eth-rpc-karura.aca-api.network',
  ACALA = 'https://eth-rpc-acala.aca-api.network',

  BSC_TESTNET = 'https://endpoints.omniatech.io/v1/bsc/testnet/public',
  FUJI = 'https://avalanche-fuji-c-chain.publicnode.com',
  GOERLI = 'https://rpc.ankr.com/eth_goerli',
  KARURA_TESTNET = 'https://eth-rpc-karura-testnet.aca-staging.network',
};

export const enum BSC_TOKEN {
  USDC = '0xB04906e95AB5D797aDA81508115611fee694c2b3',
  DAI = '0x3413a030EF81a3dD5a302F4B4D11d911e12ed337',
  USDT = '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd',
};

export const enum FUJI_TOKEN {
  USDC = '0x63A30f239DC8d1c17Bf6653a68Fc6C2F83641E6d',
};

export const WORMHOLE_GUARDIAN_RPC = {
  TESTNET: ['https://wormhole-v2-testnet-api.certus.one'],
  MAINNET: ['https://wormhole-v2-mainnet-api.certus.one'],
};

const RELAYER_BASE_URL = 'http://localhost:3111';
// const RELAYER_BASE_URL = 'https://relayer.aca-dev.network';
// const RELAYER_BASE_URL = 'https://relayer.aca-api.network';

export const RELAYER_API = {
  SHOULD_RELAY: '/shouldRelay',
  RELAY: '/relay',

  SHOULD_ROUTE_XCM: '/shouldRouteXcm',
  ROUTE_XCM: '/routeXcm',

  SHOULD_ROUTE_WORMHOLE: '/shouldRouteWormhole',
  ROUTE_WORMHOLE: '/routeWormhole',
  RELAY_AND_ROUTE: '/relayAndRoute',
  RELAY_AND_ROUTE_BATCH: '/relayAndRouteBatch',

  GET_HOMA_ROUTER_ADDR: '/shouldRouteHoma',
  ROUTE_HOMA: '/routeHoma',

  NO_ROUTE: '/noRoute',
  VERSION: '/version',
  TEST_TIMEOUT: '/testTimeout',
  HEALTH: '/health',
} as const;

export const RELAYER_URL = {
  SHOULD_RELAY: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_RELAY}`,
  RELAY: `${RELAYER_BASE_URL}${RELAYER_API.RELAY}`,

  SHOULD_ROUTE_XCM: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_ROUTE_XCM}`,
  ROUTE_XCM: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_XCM}`,

  SHOULD_ROUTE_WORMHOLE: `${RELAYER_BASE_URL}${RELAYER_API.SHOULD_ROUTE_WORMHOLE}`,
  ROUTE_WORMHOLE: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_WORMHOLE}`,
  RELAY_AND_ROUTE: `${RELAYER_BASE_URL}${RELAYER_API.RELAY_AND_ROUTE}`,
  RELAY_AND_ROUTE_BATCH: `${RELAYER_BASE_URL}${RELAYER_API.RELAY_AND_ROUTE_BATCH}`,

  GET_HOMA_ROUTER_ADDR: `${RELAYER_BASE_URL}${RELAYER_API.GET_HOMA_ROUTER_ADDR}`,
  ROUTE_HOMA: `${RELAYER_BASE_URL}${RELAYER_API.ROUTE_HOMA}`,

  NO_ROUTE: `${RELAYER_BASE_URL}${RELAYER_API.NO_ROUTE}`,
  VERSION: `${RELAYER_BASE_URL}${RELAYER_API.VERSION}`,
  TEST_TIMEOUT: `${RELAYER_BASE_URL}${RELAYER_API.TEST_TIMEOUT}`,
  HEALTH: `${RELAYER_BASE_URL}${RELAYER_API.HEALTH}`,
} as const;

/* ---------------
   thredhold amount is defined as "amount that will show on VAA"
   address should be **lower case** address to be consistent
                                                --------------- */
const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_DEV = {
  [CHAIN_ID_KARURA]: {
    // 0.1 BSC USDT => karura WUSDT 0x478dEFc2Fc2be13a505dafBDF1e5400847E2efF6
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',

    // 0.001 Goerli USDC => karura WUSDC 0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7
    '0x07865c6e87b9f70255377e024ace6630c1eaa37f': '1000',
  },
  [CHAIN_ID_ACALA]: {
    // 0.1 BSC USDT => Acala WUSDT 0xb54bA7F042DCAFba0A546e69F69E81b4F59B9C92
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',

    // 0.01 Goerli USDC => karura WUSDC 0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7
    '0x07865c6e87b9f70255377e024ace6630c1eaa37f': '10000',
  },
};

const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_PROD = {
  [CHAIN_ID_KARURA]: {
    // 10 ETH mainnet USDC => karura WUSDC 0x1F3a10587A20114EA25Ba1b388EE2dD4A337ce27
    '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': '10000000',     // 1 USDC = 10{6}

    // 10 ETH mainnet DAI  => karura WDAI 0x4bB6afB5Fa2b07a5D1c499E1c3DDb5A15e709a71
    '0x6b175474e89094c44da98b954eedeac495271d0f': '1000000000',   // 1 DAI = 10{8}

    // 1 Acala AUSD        => karura WAUSD 0xE20683ad1ED8bbeED7E1aE74Be10F19D8045B530
    '0x0000000000000000000100000000000000000001': '100000000',    // 1 AUSD = 10{8}
  },
  [CHAIN_ID_ACALA]: {
    // 10 ETH mainnet USDC => karura WUSDC 0x07DF96D1341A7d16Ba1AD431E2c847d978BC2bCe
    '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48': '10000000',     // 1 USDC = 10{6}

    // 10 ETH mainnet DAI  => karura WDAI 0x54A37A01cD75B616D63E0ab665bFfdb0143c52AE
    '0x6b175474e89094c44da98b954eedeac495271d0f': '1000000000',   // 1 DAI = 10{8}

    // 1 karura WAUSD      => Acala AUSD 0x0000000000000000000100000000000000000001
    '0x0000000000000000000100000000000000000001': '100000000',   // 1 WAUSD = 10{8}
  },
};

export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = isTestnet
  ? RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_DEV
  : RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS_PROD;

export const enum PARA_ID {
  BASILISK = '2090',
  CALAMARI = '2084',
  SHADOW = '2012',

  HYDRA = '2034',
  MANTA = '2104',
  CRUST =  '2008',
}

export const GOERLI_USDC = '0x07865c6e87b9f70255377e024ace6630c1eaa37f';
export const ZERO_ADDR = '0x0000000000000000000000000000000000000000';

const ROUTE_SUPPORTED_CHAINS_AND_ASSETS_DEV = {
  //   [HYDRA_PARA_ID]: [
  //     GOERLI_USDC,
  //   ],
  [PARA_ID.BASILISK]: [
    GOERLI_USDC,
  ],
} as const;

const ASSETS_COMMON_HYDRA = [
  ROUTER_TOKEN_INFO.weth.originAddr,
  ROUTER_TOKEN_INFO.wbtc.originAddr,
] as const;

const ASSETS_COMMON_MANTA = [
  ROUTER_TOKEN_INFO.arb.originAddr,
  ROUTER_TOKEN_INFO.ldo.originAddr,
  ROUTER_TOKEN_INFO.shib.originAddr,
  ROUTER_TOKEN_INFO.wmatic.originAddr,
  ROUTER_TOKEN_INFO.wbnb.originAddr,
  ROUTER_TOKEN_INFO.uni.originAddr,
  ROUTER_TOKEN_INFO.busd.originAddr,
  ROUTER_TOKEN_INFO.link.originAddr,
  ROUTER_TOKEN_INFO.ape.originAddr,
] as const;

export const ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD = {
  /* --------------- karura --------------- */
  [PARA_ID.BASILISK]: [
    ...ASSETS_COMMON_HYDRA,
    ROUTER_TOKEN_INFO.usdc.originAddr,
  ],
  [PARA_ID.CALAMARI]: [
    ...ASSETS_COMMON_MANTA,
    ROUTER_TOKEN_INFO.wbtc.originAddr,
    ROUTER_TOKEN_INFO.weth.originAddr,
    ROUTER_TOKEN_INFO.usdc.originAddr,
    ROUTER_TOKEN_INFO.usdt.originAddr,
  ],
  [PARA_ID.SHADOW]: [
    ROUTER_TOKEN_INFO.csm.originAddr,
  ],

  /* --------------- acala --------------- */
  [PARA_ID.HYDRA]: [
    ...ASSETS_COMMON_HYDRA,
    ROUTER_TOKEN_INFO.dai.originAddr,
    ROUTER_TOKEN_INFO.ape.originAddr,
  ],
  [PARA_ID.MANTA]: [
    ...ASSETS_COMMON_MANTA,
  ],
  [PARA_ID.CRUST]: [
    ROUTER_TOKEN_INFO.cru.originAddr,
  ],
} as const;

export const ROUTE_SUPPORTED_CHAINS_AND_ASSETS = isTestnet
  ? ROUTE_SUPPORTED_CHAINS_AND_ASSETS_DEV
  : ROUTE_SUPPORTED_CHAINS_AND_ASSETS_PROD;

export const DEST_PARA_ID_TO_ROUTER_WORMHOLE_CHAIN_ID = {
  [PARA_ID.BASILISK]: CHAIN_ID_KARURA,
  [PARA_ID.CALAMARI]: CHAIN_ID_KARURA,
  [PARA_ID.SHADOW]: CHAIN_ID_KARURA,

  [PARA_ID.HYDRA]: CHAIN_ID_ACALA,
  [PARA_ID.MANTA]: CHAIN_ID_ACALA,
  [PARA_ID.CRUST]: CHAIN_ID_ACALA,
} as const;

export const ROUTER_TOKEN_INFO_TESTNET = {
  usdc: {
    originChain: 'ETH',
    originAddr: GOERLI_USDC,
    karuraAddr: '0xE5BA1e8E6BBbdC8BbC72A58d68E74B13FcD6e4c7',
    acalaAddr: null,
    decimals: 6,
    fee: 0.04,
  },
} as const;

export const TESTNET_MODE_WARNING = `
  ----------------------------
  üî® running in testnet mode
  ‚ùå don't use it for mainnet!
  ----------------------------
`;

export const VERSION = '1.5.1';
