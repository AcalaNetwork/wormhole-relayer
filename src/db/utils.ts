import { isEvmAddress } from '@acala-network/eth-providers';
import { prisma } from './client';

export interface RouterInfo {
  params: string;       // stringified JSON params
  factoryAddr: string;
  recipient: string;
  feeAddr: string;
  routerAddr: string;
}

export const connectDb = async () => {
  try {
    await prisma.$connect();
  } catch (err) {
    console.error('❗️ db connection failed!');
    throw err;
  }
};

const _sanitizeParams = (params: string) => {
  let jsonParams: Record<string, string> = {};
  try {
    jsonParams = JSON.parse(params);
  } catch (err) {
    return params;
  }

  Object.keys(jsonParams).forEach(key => {
    if (isEvmAddress(jsonParams[key])) {
      jsonParams[key] = jsonParams[key].toLowerCase();
    }
  });

  return JSON.stringify(jsonParams);
};

// use lowercase for all addrs to be consistent
export const sanitizeRouterInfo = <T extends Partial<RouterInfo>>(info: T): T => {
  return {
    ...(info.params && { params: _sanitizeParams(info.params) }),
    ...(info.factoryAddr && { factoryAddr: info.factoryAddr.toLowerCase() }),
    ...(info.recipient && { recipient: info.recipient.toLowerCase() }),
    ...(info.feeAddr && { feeAddr: info.feeAddr.toLowerCase() }),
    ...(info.routerAddr && { routerAddr: info.routerAddr.toLowerCase() }),
  } as T;
};
