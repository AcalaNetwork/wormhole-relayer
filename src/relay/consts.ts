import {
  CHAIN_ID_KARURA,
  CHAIN_ID_ACALA,
} from '@certusone/wormhole-sdk';

// export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = {
//   [CHAIN_ID_KARURA]: {
//     '0xddb64fe46a91d46ee29420539fc25fd07c5fea3e': '10000000',   // 0.1 WETH
//     '0x2d8be6bf0baa74e0a907016679cae9190e80dd0a': '10000000',   // 0.1 ERC20
//   },
//   [CHAIN_ID_ACALA]: {
//     '0x12345': '10000',
//   },
// };

// thredhold amount is defined as "amount that will show on VAA"
// address should be lower case address to be consistent
export const RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS = {
  [CHAIN_ID_KARURA]: {
    // 0.1 BSC USDT => karura WUSDT 0x478dEFc2Fc2be13a505dafBDF1e5400847E2efF6
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',
  },
  [CHAIN_ID_ACALA]: {
    // 0.1 BSC USDT => Acala WUSDT 0xb54bA7F042DCAFba0A546e69F69E81b4F59B9C92
    '0x337610d27c682e347c9cd60bd4b3b107c9d34ddd': '10000000',

    // 0.02 BSC ETH => karura WETH ?
    '0xd66c6b4f0be8ce5b39d52e0fd1344c389929b378': '2000000',
  },
};
