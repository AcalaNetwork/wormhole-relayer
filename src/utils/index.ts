export * from './configureEnv';
export * from './logger';
export * from './relay';
export * from './route';
export * from './utils';
export * from './validate';
export * from './wormhole';
export * from './address';
export * from './error';
