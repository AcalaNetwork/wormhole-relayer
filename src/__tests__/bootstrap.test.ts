import { ADDRESSES, ROUTER_TOKEN_INFO } from '@acala-network/asset-router/dist/consts';
import { AcalaJsonRpcProvider } from '@acala-network/eth-providers';
import { ApiPromise, WsProvider } from '@polkadot/api';
import { ERC20__factory } from '@certusone/wormhole-sdk/lib/cjs/ethers-contracts';
import { FeeRegistry__factory } from '@acala-network/asset-router/dist/typechain-types';
import { LDOT } from '@acala-network/contracts/utils/AcalaTokens';
import { Wallet } from 'ethers';
import { describe, expect, it } from 'vitest';
import { parseEther, parseUnits } from 'ethers/lib/utils';
import { toHuman } from '@acala-network/asset-router/dist/utils';

import { DropAndBootstrapParams } from '../utils';
import { ETH_RPC, EUPHRATES_ADDR } from '../consts';
import {
  JITOSOL_LDOT_LP_PREDEPLOY_CODE,
  NEW_DEX_CODE,
  TEST_ADDR_RELAYER,
  TEST_ADDR_USER,
  TEST_KEY,
} from './testConsts';
import {
  api,
  expectError,
  sudoSendAndWait,
  sudoTransferToken,
  transferToken,
} from './testUtils';

const provider = new AcalaJsonRpcProvider(ETH_RPC.LOCAL);
const relayer = new Wallet(TEST_KEY.RELAYER, provider);
const user = Wallet.createRandom().connect(provider);

const JITOSOL_DECIMALS = 9;
const LDOT_DECIMALS = 10;
const JITOSOL_ADDR = ROUTER_TOKEN_INFO.jitosol.acalaAddr;

const recipient = user.address;
const boostrapAmountJitosol = '0.1';
const boostrapAmountLdot = '10';

describe('prepare', () => {
  it('open bootstrap', async () => {
    const api = await ApiPromise.create({
      provider: new WsProvider('ws://localhost:8000'),
    });

    console.log('register jitosol ...');
    let tx = api.tx.assetRegistry.registerErc20Asset(JITOSOL_ADDR, 10000);
    await sudoSendAndWait(api, tx);

    console.log('set new dex code ...');
    tx = api.tx.evm.setCode('0x0000000000000000000000000000000000000803', NEW_DEX_CODE);
    await sudoSendAndWait(api, tx);
    expect(await provider.getCode('0x0000000000000000000000000000000000000803')).to.not.eq('0x');

    console.log('deploy LDOT-jitoSOL LP token predeploy ...');
    tx = api.tx.evm.createPredeployContract('0x00000000000000000002000000000301a7fb0045', JITOSOL_LDOT_LP_PREDEPLOY_CODE, 0, 1500000, 20000, []);
    await sudoSendAndWait(api, tx);
    expect(await provider.getCode('0x00000000000000000002000000000301a7fb0045')).to.not.eq('0x');

    console.log('open euphrates pool ...');
    const evmTx = api.tx.evm.call(EUPHRATES_ADDR, '0xd914cd4b00000000000000000000000000000000000000000002000000000301a7fb0045', 0, 1000000, 100000, []);
    tx = api.tx.utility.dispatchAs({
      system: {
        Signed: '23j4ay2zBSgaSs18xstipmHBNi39W2Su9n8G89kWrz8eCe8F',
      },
    }, evmTx);
    await sudoSendAndWait(api, tx);

    console.log('list provisioning ...');
    tx = api.tx.dex.listProvisioning(
      { Token: 'LDOT' },
      { Erc20: JITOSOL_ADDR },
      1,
      1,
      parseUnits('10', 10).toBigInt(),
      parseUnits('10', 9).toBigInt(),
      0,
    );
    await sudoSendAndWait(api, tx);

    console.log('minting some jitosol for relayer ...');
    await sudoTransferToken(
      '0x335B26d05DacA31d6d921F07466e19eF8650D4Bf',
      TEST_ADDR_RELAYER,
      provider,
      JITOSOL_ADDR,
      5,
    );
  });
});


describe.concurrent('/shouldRouteDropAndBootstrap', () => {
  const testShouldRouteDropAndBootstrap = async (
    params: DropAndBootstrapParams,
    snapshotName: string,
  ) => {
    let res = await api.shouldRouteDropAndBootstrap(params);
    expect(res).toMatchSnapshot(`${snapshotName}_original`);

    // should be case insensitive
    res = await api.shouldRouteDropAndBootstrap(params);
    expect(res).toMatchSnapshot(`${snapshotName}_lowercase`);
  };

  it('when should route', async () => {
    await testShouldRouteDropAndBootstrap({
      recipient: TEST_ADDR_USER,
      gasDrop: true,
      feeToken: 'jitosol',
    }, 'jitosol_gasDrop_true');

    await testShouldRouteDropAndBootstrap({
      recipient: TEST_ADDR_USER,
      gasDrop: false,
      feeToken: 'jitosol',
    }, 'jitosol_gasDrop_false');

    await testShouldRouteDropAndBootstrap({
      recipient: TEST_ADDR_USER,
      gasDrop: false,
      feeToken: 'ldot',
    }, 'ldot_gasDrop_false');
  });

  describe('when should not route', () => {
    it('when missing params', async () => {
      try {
        await api.shouldRouteDropAndBootstrap({
          recipient,
          gasDrop: false,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['feeToken is a required field'], 400);
      }

      try {
        await api.shouldRouteDropAndBootstrap({
          recipient,
          feeToken: 'jitosol',
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['gasDrop is a required field'], 400);
      }
    });

    it('when bad params', async () => {
      try {
        await api.shouldRouteDropAndBootstrap({
          recipient,
          gasDrop: false,
          feeToken: 'xxx',
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['feeToken must be one of { jitosol, ldot }'], 400);
      }

      try {
        await api.shouldRouteDropAndBootstrap({
          recipient,
          gasDrop: true,
          feeToken: 'ldot',
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['does not support gasDrop when feeToken is ldot'], 400);
      }
    });
  });
});

describe('/routeDropAndBootstrap', () => {
  let routerAddr: string;
  const jitosol = ERC20__factory.connect(JITOSOL_ADDR, relayer);
  const ldot = ERC20__factory.connect(LDOT, relayer);

  const fetchTokenBalances = async () => {
    if (!routerAddr) throw new Error('routerAddr not set');

    const [
      userBal,
      userBalJitosol,
      relayerBalJitosol,
      routerBalJitosol,
      userBalLdot,
      relayerBalLdot,
      routerBalLdot,
    ] = await Promise.all([
      provider.getBalance(recipient),
      jitosol.balanceOf(recipient),
      jitosol.balanceOf(TEST_ADDR_RELAYER),
      jitosol.balanceOf(routerAddr),
      ldot.balanceOf(recipient),
      ldot.balanceOf(TEST_ADDR_RELAYER),
      ldot.balanceOf(routerAddr),
    ]);

    console.log({
      userBal: toHuman(userBal, 18),
      userBalJitosol: toHuman(userBalJitosol, JITOSOL_DECIMALS),
      relayerBalJitosol: toHuman(relayerBalJitosol, JITOSOL_DECIMALS),
      routerBalJitosol: toHuman(routerBalJitosol, JITOSOL_DECIMALS),
      userBalLdot: toHuman(userBalLdot, LDOT_DECIMALS),
      relayerBalLdot: toHuman(relayerBalLdot, LDOT_DECIMALS),
      routerBalLdot: toHuman(routerBalLdot, LDOT_DECIMALS),
    });

    return {
      userBal,
      userBalJitosol,
      relayerBalJitosol,
      routerBalJitosol,
      userBalLdot,
      relayerBalLdot,
      routerBalLdot,
    };
  };

  it('works with jitosol as fee token and gas drop', async () => {
    const relayerBal = await relayer.getBalance();
    expect(relayerBal.gt(parseEther('10'))).to.be.true;

    const routeArgs = {
      recipient: user.address,
      gasDrop: true,
      feeToken: 'jitosol',
    };

    const shouldRouteRes = await api.shouldRouteDropAndBootstrap(routeArgs);
    ({ routerAddr } = shouldRouteRes.data);
    console.log({ routerAddr });

    // make sure user has enough token and ACA to transfer to router
    console.log('refilling ACA for user ...');
    await (await relayer.sendTransaction({
      to: recipient,
      value: parseEther('3'),
    })).wait();

    const bal = await fetchTokenBalances();
    const refillAmount = parseUnits(boostrapAmountJitosol, JITOSOL_DECIMALS);
    if (bal.userBalJitosol.lt(refillAmount.mul(2))) {
      if (bal.relayerBalJitosol.lt(refillAmount.mul(2))) {
        throw new Error('both relayer and user do not have enough jitosol to transfer to router!');
      }

      console.log('refilling token for user ...');
      await (await jitosol.transfer(recipient, refillAmount.mul(2))).wait();
    }

    console.log('transferring token to router ...');
    await transferToken(routerAddr, user, JITOSOL_ADDR, Number(boostrapAmountJitosol));

    const bal0 = await fetchTokenBalances();

    console.log('routing ...');
    let routeRes = await api.routeDropAndBootstrap(routeArgs);
    let txHash = routeRes.data;
    console.log(`route finished! txHash: ${txHash}`);

    const bal1 = await fetchTokenBalances();

    // router should NOT be destroyed
    let routerCode = await provider.getCode(routerAddr);
    expect(routerCode.length).to.be.greaterThan(100);
    expect(bal1.routerBalJitosol.toNumber()).to.eq(0);
    expect(bal1.routerBalLdot.toNumber()).to.eq(0);

    // relayer should receive jitosol fee
    const routingFee = await FeeRegistry__factory.connect(ADDRESSES.ACALA.feeAddr, provider)
      .getFee(JITOSOL_ADDR);

    // first route should perform swap + ACA drop
    const swapFee = parseUnits('0.0035', 9);
    expect(bal1.relayerBalJitosol.sub(bal0.relayerBalJitosol).toBigInt())
      .to.eq(routingFee.add(swapFee).toBigInt());
    expect(bal1.relayerBalLdot.sub(bal0.relayerBalLdot).toBigInt()).to.eq(0n);

    // user should receive 3 ACA drop
    expect(bal1.userBal.sub(bal0.userBal).toBigInt()).to.eq(parseEther('3').toBigInt());

    console.log('------------------------------------ route 2 ------------------------------------');
    console.log('transferring token to router ...');
    await transferToken(routerAddr, user, JITOSOL_ADDR, Number(boostrapAmountJitosol));

    const bal2 = await fetchTokenBalances();

    console.log('routing ...');
    routeRes = await api.routeDropAndBootstrap(routeArgs);
    txHash = routeRes.data;
    console.log(`route finished! txHash: ${txHash}`);

    const bal3 = await fetchTokenBalances();

    // router should NOT be destroyed
    routerCode = await provider.getCode(routerAddr);
    expect(routerCode.length).to.be.greaterThan(100);
    expect(bal3.routerBalJitosol.toNumber()).to.eq(0);
    expect(bal3.routerBalLdot.toNumber()).to.eq(0);

    // second route should only route
    expect(bal3.relayerBalJitosol.sub(bal2.relayerBalJitosol).toBigInt()).to.eq(routingFee.toBigInt());
    expect(bal3.relayerBalLdot.sub(bal2.relayerBalLdot).toBigInt()).to.eq(0n);

    // user should NOT receive 3 ACA drop
    expect(bal3.userBal.sub(bal2.userBal).toBigInt()).to.eq(0n);
  });

  it('works with jitosol as fee token and no gas drop', async () => {
    const relayerBal = await relayer.getBalance();
    expect(relayerBal.gt(parseEther('10'))).to.be.true;

    const routeArgs = {
      recipient: user.address,
      gasDrop: false,
      feeToken: 'jitosol',
    };

    const shouldRouteRes = await api.shouldRouteDropAndBootstrap(routeArgs);
    ({ routerAddr } = shouldRouteRes.data);
    console.log({ routerAddr });

    // make sure user has enough token and ACA to transfer to router
    console.log('refilling ACA for user ...');
    await (await relayer.sendTransaction({
      to: recipient,
      value: parseEther('3'),
    })).wait();

    const bal = await fetchTokenBalances();
    const refillAmount = parseUnits(boostrapAmountJitosol, JITOSOL_DECIMALS);
    if (bal.userBalJitosol.lt(refillAmount.mul(2))) {
      if (bal.relayerBalJitosol.lt(refillAmount.mul(2))) {
        throw new Error('both relayer and user do not have enough jitosol to transfer to router!');
      }

      console.log('refilling token for user ...');
      await (await jitosol.transfer(recipient, refillAmount.mul(2))).wait();
    }

    console.log('transferring token to router ...');
    await transferToken(routerAddr, user, JITOSOL_ADDR, Number(boostrapAmountJitosol));

    const bal0 = await fetchTokenBalances();

    console.log('routing ...');
    let routeRes = await api.routeDropAndBootstrap(routeArgs);
    let txHash = routeRes.data;
    console.log(`route finished! txHash: ${txHash}`);

    const bal1 = await fetchTokenBalances();

    // router should NOT be destroyed
    let routerCode = await provider.getCode(routerAddr);
    expect(routerCode.length).to.be.greaterThan(100);
    expect(bal1.routerBalJitosol.toNumber()).to.eq(0);
    expect(bal1.routerBalLdot.toNumber()).to.eq(0);

    // relayer should receive jitosol fee but no swap fee
    const routingFee = await FeeRegistry__factory.connect(ADDRESSES.ACALA.feeAddr, provider)
      .getFee(JITOSOL_ADDR);
    expect(bal1.relayerBalJitosol.sub(bal0.relayerBalJitosol).toBigInt()).to.eq(routingFee.toBigInt());
    expect(bal1.relayerBalLdot.sub(bal0.relayerBalLdot).toBigInt()).to.eq(0n);

    // user should NOT receive 3 ACA drop
    expect(bal1.userBal.sub(bal0.userBal).toBigInt()).to.eq(0n);

    console.log('------------------------------------ route 2 ------------------------------------');
    console.log('transferring token to router ...');
    await transferToken(routerAddr, user, JITOSOL_ADDR, Number(boostrapAmountJitosol));

    const bal2 = await fetchTokenBalances();

    console.log('routing ...');
    routeRes = await api.routeDropAndBootstrap(routeArgs);
    txHash = routeRes.data;
    console.log(`route finished! txHash: ${txHash}`);

    const bal3 = await fetchTokenBalances();

    // router should NOT be destroyed
    routerCode = await provider.getCode(routerAddr);
    expect(routerCode.length).to.be.greaterThan(100);
    expect(bal3.routerBalJitosol.toNumber()).to.eq(0);
    expect(bal3.routerBalLdot.toNumber()).to.eq(0);

    // second route should be the same as the first one
    expect(bal3.relayerBalJitosol.sub(bal2.relayerBalJitosol).toBigInt()).to.eq(routingFee.toBigInt());
    expect(bal3.relayerBalLdot.sub(bal2.relayerBalLdot).toBigInt()).to.eq(0n);

    // user should NOT receive 3 ACA drop
    expect(bal3.userBal.sub(bal2.userBal).toBigInt()).to.eq(0n);
  });
});
