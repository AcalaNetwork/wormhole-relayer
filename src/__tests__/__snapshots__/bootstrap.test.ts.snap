// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/shouldRouteDropAndBootstrap > when should route > jitosol_gasDrop_false_lowercase 1`] = `
{
  "data": {
    "routerAddr": "0x7dD4A59B255e13F2eb5cF0297574e5fb9e89576b",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteDropAndBootstrap > when should route > jitosol_gasDrop_false_original 1`] = `
{
  "data": {
    "routerAddr": "0x7dD4A59B255e13F2eb5cF0297574e5fb9e89576b",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteDropAndBootstrap > when should route > jitosol_gasDrop_true_lowercase 1`] = `
{
  "data": {
    "routerAddr": "0x9A1100129688460930096d7E1204A2989321B113",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteDropAndBootstrap > when should route > jitosol_gasDrop_true_original 1`] = `
{
  "data": {
    "routerAddr": "0x9A1100129688460930096d7E1204A2989321B113",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteDropAndBootstrap > when should route > ldot_gasDrop_false_lowercase 1`] = `
{
  "data": {
    "routerAddr": "0x6046DD5F660818f9D64Af4536A061D8C3BCE01b9",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteDropAndBootstrap > when should route > ldot_gasDrop_false_original 1`] = `
{
  "data": {
    "routerAddr": "0x6046DD5F660818f9D64Af4536A061D8C3BCE01b9",
    "shouldRoute": true,
  },
}
`;
